mdnsd/mdnsd.8
.\"
.\" Copyright (c) 2010, 2011, Christiano F. Haesbaert <haesbaert@haesbaert.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: Feb 06 2011 $
.Dt MDNSD 8
.Os
.Sh NAME
.Nm mdnsd
.Nd Multicast DNS/DNS-SD daemon
.Sh SYNOPSIS
.Nm
.Op Fl dvw
.Ar ifname
.Op Ar ifnames ...
.Sh DESCRIPTION
.Nm
is a Multicast Domain Name System
.Pq mDNS
daemon which acts as the host mDNS querier and responder.
.Nm
supports both raw mDNS as well as DNS-SD (Service Discovery) as described in
the mDNS and DNS-SD RFCs.
.Pp
mDNS is a way to perform DNS-like operations via multicast in the local link,
there is no hierarchy or multiple domains as in conventional unicast DNS.
mDNS provides a way for hosts to co-operate and maintain a cache name database
which can then be used to resolve local host names without the need of a
central DNS server.
.Pp
DNS-SD is a convention on some names used in mDNS to provide hosts with
Service Discovery capabilities.
A host can publish a service of any type, be it a HTTP server, NTP server, a
printer and so on, this services can then be browsed and resolved by the other
hosts in the local network.
.Pp
There are basically two roles in a mDNS environment, the Querier, and the
Responder.
The Querier is the entity responsible for sending questions and mDNS requests
in the local link, it can't be done as it is in libc, where each process does
its own lookup, there must be something centralizing all the requests as there
are various complications implied: cache, timers and so on.
.Pp
The Responder is the entity responsible for answering those queries, there
should be only one responder per host.
Both roles are performed by mDNS.
mDNS operates on the link-local multicast address 224.0.0.251 under UDP
port 5353.
There are no multiple domains in mDNS as in unicast DNS, the .local domain
name is the single mDNS domain name and it's where all the queries and answers
take place.
.Pp
There are three basic types of mDNS question, in which
.Nm
uses two of them.
The One-Shot Query, which resembles unicast DNS, where a single question is
sent and an answer is expected, if no answer is received it means no one can
answer that question.
This question is used for simple lookups.
.Pp
Continuous Multicast Query is a more complex way of querying, the querier will
send the same question multiple times, doubling the interval between each
time, multiple answers may be received, it's used as a way for monitoring the
Resource Records of the network.
This question is mainly used by network browsing in DNS-SD, where a question
for a type of service may enumerate one or more instances, for example, if
browsing for the HTTP servers, there may be one or more servers (instances).
To diminish the volume of redundant answers, a feature called Known Answer
Suppression is present, in which the querier when performing a Continuous
Multicast Query places all the previous known answers in the additional
section of the mDNS packet, thus, any answer that would be given which is
already in the additional section, is suppressed.
.Pp
There are two type of Resource Records, Unique and Shared.
.Pp
Unique records are the ones which there may be only one answer for it in the
local name, the A, PTR and HINFO under the hostname.local name are examples of
Unique records, it would be strange if two hosts would answer an address for
the same foobar.local.
All Unique records must be Probed to verify its uniqueness, if a conflict is
found, another name must be chosen (Unimplemented).
.Pp
A Shared record is used for PTR records in DNS-SD, a host may have as many
answers as necessary for a shared record, it's used only in network browsing,
where there may be multiple instances of the same service.
.Pp
To access the mDNS services, a libmdns library will be provided in the near
future, programs will then, be able to link with libmdns and publish its own
services though mDNS.
By now, only mdnsctl(8) is provided which is a command line interface to the
daemon in the same fashion as ripctl(8) and ospfctl(8).
.Pp
When
.Nm
starts up, it probes for its hostname (fetched from /etc/myname), if there
isn't a conflict, it publishes an A and a PTR record for itself, both records
will be under the .local domain, which is the mDNS single domain name.
All the other domain names in /etc/myname will be stripped, therefore
foo.bar.midearth becomes foo.local, which can be resolved through mDNS.
If a conflict is found, then, there is another foo.local in the network and
conflict resolution takes place.
.Pp
If
Fl w
was not specified,
.Nm
will also publish a Workstation service, this service has no data itself, it's
used to state that the host is up, it can be used for example, to browse every
powered host on the local network.
.Pp
.Nm
supports multiple interfaces, the interfaces used must be specified as the
arguments.
.Pp
.Nm
must be started as root and upon start up it will drop privileges, change it's
euid/egid to _mdnsd and chroot.
Therefore make sure you have user and group _mdnsd created.
.Pp
mDNS operations can be done with the
.Xr mdnsctl 8
utility.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
Do not daemonize.
If this option is specified, debugging will be enabled,
.Nm
will run in the foreground and log to
.Em stderr .
.It Fl w
Do not publish a workstation service on startup.
.It Fl v
Print version and exit.
.El
.Sh FILES
.Bl -tag -width "/var/run/mdnsd.sockXX" -compact
.It /var/run/mdnsd.sock
.Ux Ns -domain
socket used for communication with
.Xr mdnsctl 8 .
.El
.Sh SEE ALSO
.Xr mdnsctl 8
.Rs
.%A S. Cheshire
.%A M. Krochmal
.%D February 2013
.%R RFC 6762
.%T Multicast DNS
.Re
.Rs
.%A S. Cheshire
.%A M. Krochmal
.%D February 2013
.%R RFC 6763
.%T DNS-Based Service Discovery
.Re
.Sh LICENSE
.Nm
is released under the ISC license.
.Sh HISTORY
The
.Nm
program version 0.1 was released in 13 February 2011.
.Sh AUTHORS
.An Christiano Farina Haesbaert Aq Mt haesbaert@haesbaert.org
.Sh BUGS
No proper error return in mdnsl.c.


mdnsd/mdnsd.c
/*
 * Copyright (c) 2010-2014 Christiano F. Haesbaert <haesbaert@haesbaert.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/utsname.h>
#include <sys/wait.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <err.h>
#include <event.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>

#include "version.h"
#include "mdnsd.h"
#include "mdns.h"
#include "log.h"
#include "control.h"

__dead void	usage(void);
__dead void	display_version(void);
void		mdnsd_sig_handler(int, short, void *);
void		conf_init_ifaces(int, char *[]);
void		mdnsd_shutdown(void);
int		mdns_sock(struct sockaddr *);
void		fetchmyname(char [MAXHOSTNAMELEN]);
void		fetchhinfo(struct hinfo *);

struct mdnsd_conf	*conf = NULL;
extern char		*malloc_options;

__dead void
usage(void)
{
	extern char	*__progname;

	fprintf(stderr, "usage: %s [-46dw] ifname [ifnames...]\n",
	    __progname);
	fprintf(stderr, "usage: %s -v\n", __progname);
	exit(1);
}

__dead void
display_version(void)
{
	printf("OpenMdns Daemon %s\n", MDNS_VERSION);
	printf("Copyright (C) 2010-2014 Christiano F. Haesbaert\n");

	exit(0);
}

void
conf_init_ifaces(int argc, char *argv[])
{
	int		 found = 0;
	int		 i;
	struct iface	*iface;

	for (i = 0; i < argc; i++) {
		iface = if_new(argv[i]);
		if (iface == NULL)
			continue;
		found++;
		LIST_INSERT_HEAD(&conf->iface_list, iface, entry);
	}

	if (!found)
		fatal("Couldn't find any interface");

	LIST_FOREACH(iface, &conf->iface_list, entry)
		log_debug("using iface %s index %u", iface->name, iface->ifindex);
}

/* ARGSUSED */
void
mdnsd_sig_handler(int sig, short event, void *arg)
{
	/*
	 * signal handler rules don't apply, libevent decouples for us
	 */

	switch (sig) {
	case SIGTERM:
	case SIGINT:
		mdnsd_shutdown();
		break;		/* NOTREACHED */
	case SIGHUP:
		log_debug("got SIGHUP");
		/* reconfigure */
		/* ... */
		break;
	default:
		fatalx("unexpected signal");
		/* NOTREACHED */
	}
}

void
mdnsd_shutdown(void)
{
	struct iface	*iface;
	struct pge	*pge;

	/*
	 * Send goodbye RR for all published records.
	 */
	while ((pge = TAILQ_FIRST(&pge_queue)) != NULL)
		pge_kill(pge);

	while ((iface = LIST_FIRST(&conf->iface_list)) != NULL) {
		LIST_REMOVE(iface, entry);
		free(iface);
	}

	kev_cleanup();
	control_cleanup();
	free(conf);

	log_info("terminating");
	exit(0);
}


int
mdns_sock(struct sockaddr *addr)
{
	int sock;
	struct sockaddr_in *sa4;
	struct sockaddr_in6 *sa6;
	char addrstr[INET6_ADDRSTRLEN];

	switch(addr->sa_family) {
	case AF_INET:
		sa4 = (struct sockaddr_in*)addr;
		sa4->sin_len = sizeof(struct sockaddr_in);
		sa4->sin_port = htons(MDNS_PORT);
		sa4->sin_addr.s_addr = INADDR_ANY;
		break;
	case AF_INET6:
		sa6 = (struct sockaddr_in6*)addr;
		sa6->sin6_len = sizeof(struct sockaddr_in6);
		sa6->sin6_port = htons(MDNS_PORT);
		memcpy(&sa6->sin6_addr, &in6addr_any, sizeof(struct in6_addr));
		break;
	default:
		fatal("mdns_sock AF not supported");
	}

	if ((sock = socket(addr->sa_family, SOCK_DGRAM, IPPROTO_UDP)) == -1)
		fatal("socket");

	if (bind(sock, addr, addr->sa_len) == -1)
		fatal("bind");

	/* XXX stupid patchup */
	switch(addr->sa_family) {
	case AF_INET:
		sa4->sin_addr.s_addr = MDNS_INADDR;
		break;
	case AF_INET6:
		memcpy(&sa6->sin6_addr, &mdns_in6addr, sizeof(struct in6_addr));
		break;
	}

	if (if_set_opt(addr->sa_family, sock) == -1)
		fatal("if_set_opt");

	if (if_set_mcast_ttl(addr->sa_family, sock, MDNS_TTL) == -1)
		fatal("if_set_mcast_ttl");

	if (if_set_mcast_loop(addr->sa_family, sock) == -1)
		fatal("if_set_mcast_loop");

	if_set_recvbuf(sock);

	log_debug("mdns sock bound to %s:%u", satop(addr, addrstr), MDNS_PORT);

	return (sock);
}

void
fetchmyname(char myname[MAXHOSTNAMELEN])
{
	char	*end;

	if (gethostname(myname, MAXHOSTNAMELEN) == -1)
		fatal("gethostname");
	end = strchr(myname, '.');
	if (end != NULL)
		*end = '\0';	/* use short hostnames */
	if (strlcat(myname, ".local", MAXHOSTNAMELEN) >= MAXHOSTNAMELEN)
		errx(1, "hostname too long %s", myname);
}

void
fetchhinfo(struct hinfo *hi)
{
	struct utsname	utsname;

	if (uname(&utsname) == -1)
		fatal("uname");
	bzero(hi, sizeof(*hi));
	strlcpy(hi->cpu, utsname.machine, sizeof(hi->cpu));
	snprintf(hi->os, sizeof(hi->os), "%s %s", utsname.sysname,
	    utsname.release);
}

int
main(int argc, char *argv[])
{
	int		 ch;
	int		 debug, no_workstation;
	sa_family_t	 addrpref;
	struct passwd	*pw;
	struct iface	*iface;
	struct event	 ev_sigint, ev_sigterm, ev_sighup;

	debug = no_workstation = addrpref = 0;
	/*
	 * XXX Carefull not to call anything that would malloc prior to setting
	 * malloc_options, malloc will disregard malloc_options after the first
	 * call.
	 */
	while ((ch = getopt(argc, argv, "46dvw")) != -1) {
		switch (ch) {
		case '4':
			addrpref = AF_INET;
			break;
		case '6':
			addrpref = AF_INET6;
			break;
		case 'd':
			debug = 1;
			malloc_options = "AFGJPX";
			break;
		case 'v':
			display_version();
			break;	/* NOTREACHED */
		case 'w':
			no_workstation = 1;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}

	log_init(1);	/* log to stderr until daemonized */

	argc -= optind;
	argv += optind;

	if (!argc)
		usage();

	/* check for root privileges */
	if (geteuid())
		errx(1, "need root privileges");

	/* check for mdnsd user */
	if ((pw = getpwnam(MDNSD_USER)) == NULL)
		fatal("getpwnam, make sure you have user and group _mdnsd");

	log_init(debug);

	if (!debug)
		daemon(1, 0);

	log_info("startup");

	/* init control before chroot */
	if (control_init() == -1)
		fatalx("control socket setup failed");

	/* chroot */
	if (chroot(pw->pw_dir) == -1)
		fatal("chroot");
	if (chdir("/") == -1)
		fatal("chdir(\"/\")");

	/* show who we are */
	setproctitle("mdnsd");

	/* drop privileges */
	if (setgroups(1, &pw->pw_gid) ||
	    setresgid(pw->pw_gid, pw->pw_gid, pw->pw_gid) ||
	    setresuid(pw->pw_uid, pw->pw_uid, pw->pw_uid))
		fatal("error droping privileges");

	/* init libevent */
	event_init();

	/* setup signals */
	signal_set(&ev_sigint, SIGINT, mdnsd_sig_handler, NULL);
	signal_set(&ev_sigterm, SIGTERM, mdnsd_sig_handler, NULL);
	signal_set(&ev_sighup, SIGHUP, mdnsd_sig_handler, NULL);
	signal_add(&ev_sigint, NULL);
	signal_add(&ev_sigterm, NULL);
	signal_add(&ev_sighup, NULL);
	signal(SIGPIPE, SIG_IGN);

	/* init conf */
	if ((conf = calloc(1, sizeof(*conf))) == NULL)
		fatal("calloc");
	fetchmyname(conf->myname);
	fetchhinfo(&conf->hi);
	LIST_INIT(&conf->iface_list);
	conf->no_workstation = no_workstation;

	/* init RR cache */
	cache_init();

	/* init publish queues */
	pg_init();

	/* init interfaces and names */
	conf_init_ifaces(argc, argv);

	/* Create primary pge */
	pge_initprimary();

	/* init some packet internals */
	packet_init();

	/* init querier */
	query_init();

	/* listen to kernel interface events */
	kev_init();

	/* create mdns socket */
	if (addrpref == AF_UNSPEC || addrpref == AF_INET) {
		conf->udp4.ss.ss_family = AF_INET;
		conf->udp4.fd = mdns_sock(sstosa(&conf->udp4.ss));

		/* setup mdns events */
		event_set(&conf->udp4.ev, conf->udp4.fd, EV_READ|EV_PERSIST,
		    recv_packet, NULL);
		event_add(&conf->udp4.ev, NULL);
	}

	if (addrpref == AF_UNSPEC || addrpref == AF_INET6) {
		conf->udp6.ss.ss_family = AF_INET6;
		conf->udp6.fd = mdns_sock(sstosa(&conf->udp6.ss));

		/* setup mdns events */
		event_set(&conf->udp6.ev, conf->udp6.fd, EV_READ|EV_PERSIST,
		    recv_packet, NULL);
		event_add(&conf->udp6.ev, NULL);
	}

	/* start interfaces */
	LIST_FOREACH(iface, &conf->iface_list, entry) {
		if (if_fsm(iface, IF_EVT_UP))
			log_warnx("error starting interface %s", iface->name);
	}

	/* listen on mdns control socket */
	TAILQ_INIT(&ctl_conns);
	control_listen();

	/* parent mainloop */
	event_dispatch();

	return (0);
}

void
imsg_event_add(struct imsgev *iev)
{
	if (iev->handler == NULL) {
		imsg_flush(&iev->ibuf);
		return;
	}

	iev->events = EV_READ;
	if (iev->ibuf.w.queued)
		iev->events |= EV_WRITE;

	event_del(&iev->ev);
	event_set(&iev->ev, iev->ibuf.fd, iev->events, iev->handler, iev);
	event_add(&iev->ev, NULL);
}

int
mdnsd_imsg_compose_ctl(struct ctl_conn *c, u_int16_t type,
    void *data, u_int16_t datalen)
{
	return (imsg_compose_event(&c->iev, type, 0, 0, -1, data, datalen));
}

int
imsg_compose_event(struct imsgev *iev, u_int16_t type,
    u_int32_t peerid, pid_t pid, int fd, void *data, u_int16_t datalen)
{
	int	ret;

	if ((ret = imsg_compose(&iev->ibuf, type, peerid,
	    pid, fd, data, datalen)) != -1)
		imsg_event_add(iev);
	return (ret);
}

int
peersuser(int fd)
{
	uid_t	euid;

	if (getpeereid(fd, &euid, NULL) == -1)
		fatal("getpeereid");
	return (euid == 0);
}
